src/os_auth/auth.c

```diff
@@ -164,62 +166,88 @@ w_err_t w_auth_parse_data(const char* buf, char *response,const char *authpass,
+        }
+    }
+
+    /* Replace the agent */
+    minfo("Removing old agent '%s'.", key->id);
+    add_remove(key);
+    OS_DeleteKey(&keys, key->id, 0);
+    return OS_SUCCESS;
+}
+
+w_err_t w_auth_validate_data(char *response,
+                             const char *ip,
+                             const char *agentname,
+                             const char *groups,
+                             const char *key_hash){
+    int index = 0;
+
+    /* Validate the group(s) name(s) */
     if (groups){
         if (OS_SUCCESS != w_auth_validate_groups(groups, response)){
             return OS_INVALID;
         }
     }
 
     /* Check for duplicated IP */
-    if (strcmp(ip, "any") != 0 ) {
-        if (index = OS_IsAllowedIP(&keys, ip), index >= 0) {
-            if (config.flags.force_insert && (antiquity = OS_AgentAntiquity(keys.keyentries[index]->name, keys.keyentries[index]->ip->ip), antiquity >= config.force_time || antiquity < 0)) {
-                id_exist = keys.keyentries[index]->id;
-                minfo("Duplicated IP '%s' (%s). Removing old agent.", ip, id_exist);
-
-                add_remove(keys.keyentries[index]);
-                OS_DeleteKey(&keys, id_exist, 0);
-            } else {
-                merror("Duplicated IP %s", ip);
-                snprintf(response, 2048, "ERROR: Duplicated IP: %s", ip);
-                return OS_INVALID;
-            }
+    if (strcmp(ip, "any") != 0 && (index = OS_IsAllowedIP(&keys, ip), index >= 0)) {
+        minfo("Duplicated IP '%s' (%s).", ip, keys.keyentries[index]->id);
+        if (OS_SUCCESS != w_auth_replace_agent(keys.keyentries[index], key_hash)) {
+            snprintf(response, 2048, "ERROR: Duplicated IP: %s", ip);
+            return OS_INVALID;
         }
     }
 
     /* Check whether the agent name is the same as the manager */
-
     if (!strcmp(agentname, shost)) {
         merror("Invalid agent name %s (same as manager)", agentname);
         snprintf(response, 2048, "ERROR: Invalid agent name: %s", agentname);
         return OS_INVALID;
     }
 
-    /* Check for duplicated names */
-
+    /* Check for duplicated name */
     if (index = OS_IsAllowedName(&keys, agentname), index >= 0) {
-        if (config.flags.force_insert && (antiquity = OS_AgentAntiquity(keys.keyentries[index]->name, keys.keyentries[index]->ip->ip), antiquity >= config.force_time || antiquity < 0)) {
-            id_exist = keys.keyentries[index]->id;
-            minfo("Duplicated name '%s' (%s). Removing old agent.", agentname, id_exist);
-
-            add_remove(keys.keyentries[index]);
-            OS_DeleteKey(&keys, id_exist, 0);
-        } else {
-            merror("Invalid agent name %s (duplicated)", agentname);
+        minfo("Duplicated name '%s' (%s).", agentname, keys.keyentries[index]->id);
+        if (OS_SUCCESS != w_auth_replace_agent(keys.keyentries[index], key_hash)) {
             snprintf(response, 2048, "ERROR: Duplicated agent name: %s", agentname);
```

Comment:
This response message is there before this PR, but maybe we can update the error message.
`w_auth_replace_agent` may fail for the agent's antiquity or the existing key, and the cause of the fail isn't received by the agent.

Reply:
I agree, that improvement shouldn't be included in this PR.